src/treemodel/operation/attributeoperation.js

```diff
@@ -87,13 +88,22 @@ export default class AttributeOperation extends Operation {
 			 */
 			throw new CKEditorError(
 				'operation-attribute-different-keys: Old and new attributes should have the same keys.',
-				{ oldAttr: oldAttr, newAttr: newAttr } );
+				{ oldAttr: oldAttr, newAttr: newAttr }
+			);
 		}
 
+		// Split text nodes if needed, then get the nodes in the range and convert it to node list. It will be easier to operate.
+		this.range.start.parent._children._splitNodeAt( this.range.start.offset );
+		this.range.end.parent._children._splitNodeAt( this.range.end.offset );
+
 		// Remove or change.
 		if ( oldAttr !== null ) {
-			for ( let node of this.range.getAllNodes() ) {
-				if ( !node.hasAttr( oldAttr ) ) {
+			for ( let node of this.range.getAllNodes( true ) ) {
+				if ( node instanceof TextFragment ) {
+					node = node.first._nodeListText;
```

Comment:
We could get back to the point where attribute methods were a part of given class not `AttributeList` because there are more and more advantages of it. Then it would be less hacky as we may introduce special attributes methods for every class. I am for doing it in another task.